zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java

```diff
@@ -0,0 +1,135 @@
+ * be cached inside this helper object. This helper object watches the clientUri-domain ZNodes and
+ * updates the internal Map accordingly.
+ *
+ * The following illustrates the ZNode hierarchy:
+ * . (root)
+ * └── _CLIENT_URI_DOMAIN_MAPPING (mapping root path)
+ *     ├── bar (application domain)
+ *     │   ├── bar0 (client URI)
+ *     │   └── bar1 (client URI)
+ *     └── foo (application domain)
+ *         ├── foo1 (client URI)
+ *         ├── foo2 (client URI)
+ *         └── foo3 (client URI)
+ *
+ * Note: It is not expected that there would be too many distinct client URIs so as to overwhelm
+ * heap usage.
+ */
+public class ZkClientUriDomainMappingHelper implements Watcher, ClientUriDomainMappingHelper {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ZkClientUriDomainMappingHelper.class);
+
+  private static final String CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH =
+      "zookeeper.znode.groupacl.clientUriDomainMappingRootPath";
+
+  private final ZooKeeperServer zks;
+  private final String rootPath;
+  private final Map<String, Set<String>> clientUriToDomainNames = new HashMap<>();
+
+  public ZkClientUriDomainMappingHelper(ZooKeeperServer zks) {
+    this.zks = zks;
+
+    this.rootPath = System.getProperty(CLIENT_URI_DOMAIN_MAPPING_ROOT_PATH);
+    if (rootPath == null) {
+      throw new IllegalStateException(
+          "ZkClientUriDomainMappingHelper::ClientUriDomainMapping root path config is not set!");
+    }
+
+    if (zks.getZKDatabase().getNode(rootPath) == null) {
+      throw new IllegalStateException(
+          "ZkClientUriDomainMappingHelper::ClientUriDomainMapping root path does not exist!");
+    }
+
+    addWatches();
+    parseZNodeMapping();
+  }
+
+  /**
+   * Install a persistent recursive watch on the root path.
+   */
+  private void addWatches() {
+    zks.getZKDatabase().addWatch(rootPath, this, ZooDefs.AddWatchModes.persistentRecursive);
+  }
+
+  /**
+   * Read ZNodes under the root path and populates clientUriToDomainNames.
+   * Note: this is not thread-safe nor atomic; however, we do not need such strong guarantee with
+   * this read operation.
+   *
+   * Also, note that this is a purely in-memory operation, so re-parsing the entire tree should not
+   * be a big overhead considering how infrequently the mapping is supposed to be changed.
+   */
+  private void parseZNodeMapping() {
+    clientUriToDomainNames.clear();
+    try {
+      List<String> domainNames = zks.getZKDatabase().getChildren(rootPath, null, null);
+      domainNames.forEach(domainName -> {
+        try {
+          List<String> clientUris =
+              zks.getZKDatabase().getChildren(rootPath + "/" + domainName, null, null);
+          clientUris.forEach(
+              clientUri -> clientUriToDomainNames.computeIfAbsent(clientUri, k -> new HashSet<>())
+                  .add(domainName));
+        } catch (KeeperException.NoNodeException e) {
```

Comment:
I think we should not catch the exception so the down stream knows there is an error

Reply:
This is not a failure condition. Having no ZNode in this path is a valid state to be in - for example, the user hasn't finished configuring the znodes. This is why we catch the error. 

Let me know if that makes sense.